File: RaAndroidNotesApp/app/src/main/java/com/ralph/notes/BlogRepository.java
package com.ralph.notes;

import android.content.Context;

import org.json.JSONObject;
import org.json.JSONArray;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.net.HttpURLConnection;
import java.net.ConnectException;
import java.net.SocketTimeoutException;
import java.net.UnknownHostException;
import java.net.URLEncoder;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class BlogRepository {
    public static final String PUBLIC_DATA_URL = "https://galphrui.github.io/ralphrong/data/posts.json";
    public static final String WORKER_BASE_URL = "https://ralphrong-blog-admin.ralphrong.workers.dev";

    private final Context context;
    private final CacheStore cacheStore;

    public BlogRepository(Context context) {
        this.context = context.getApplicationContext();
        this.cacheStore = new CacheStore(this.context);
    }

    public BlogData fetchPublicData() throws Exception {
        JSONObject json;
        BlogData data;
        try {
            String body = request("GET", PUBLIC_DATA_URL + "?t=" + System.currentTimeMillis(), null, null);
            json = new JSONObject(body);
            cacheStore.saveBlog(json);
            data = JsonMapper.parseBlog(json);
            data.offlineMode = false;
            data.fromCache = false;
            data.sourceMessage = "公网数据";
        } catch (Exception error) {
            json = cacheStore.loadBlog();
            if (json != null) {
                data = JsonMapper.parseBlog(json);
                data.offlineMode = true;
                data.fromCache = true;
                data.sourceMessage = "本地缓存";
            } else {
                String body = readAll(context.getAssets().open("offline-posts.json"));
                data = JsonMapper.parseBlog(new JSONObject(body));
                data.offlineMode = true;
                data.fromCache = false;
                data.sourceMessage = "内置离线数据";
            }
        }
        return data;
    }

    public VisitStats recordVisit(String visitorId) throws Exception {
        try {
            JSONObject body = new JSONObject();
            body.put("visitorId", visitorId);
            JSONObject response = new JSONObject(request("POST", WORKER_BASE_URL + "/api/visits", body, null));
            VisitStats stats = statsFromResponse(response);
            stats.fromCache = false;
            stats.sourceMessage = "公网统计";
            cacheStore.saveStats(stats);
            return stats;
        } catch (Exception error) {
            VisitStats cached = cacheStore.loadStats();
            if (cached != null) return cached;
            throw error;
        }
    }

    public VisitStats fetchVisitStats() throws Exception {
        try {
            JSONObject response = new JSONObject(request("GET", WORKER_BASE_URL + "/api/visits", null, null));
            VisitStats stats = statsFromResponse(response);
            stats.fromCache = false;
            stats.sourceMessage = "公网统计";
            cacheStore.saveStats(stats);
            return stats;
        } catch (Exception error) {
            VisitStats cached = cacheStore.loadStats();
            if (cached != null) return cached;
            throw error;
        }
    }

    public List<GuestMessage> fetchMessages() throws Exception {
        return fetchMessages("");
    }

    public List<GuestMessage> fetchMessages(String postSlug) throws Exception {
        String suffix = postSlug == null || postSlug.isEmpty() ? "" : "?postSlug=" + URLEncoder.encode(postSlug, "UTF-8");
        JSONObject response = new JSONObject(request("GET", WORKER_BASE_URL + "/api/messages" + suffix, null, null));
        if (!response.optBoolean("ok")) throw new IllegalStateException(response.optString("error", "读取留言失败"));
        return messagesFromResponse(response);
    }

    public List<GuestMessage> createMessage(String name, String message) throws Exception {
        return createMessage(name, message, "");
    }

    public List<GuestMessage> createMessage(String name, String message, String postSlug) throws Exception {
        JSONObject body = new JSONObject();
        body.put("name", name);
        body.put("message", message);
        body.put("postSlug", postSlug == null ? "" : postSlug);
        JSONObject response = new JSONObject(request("POST", WORKER_BASE_URL + "/api/messages", body, null));
        if (!response.optBoolean("ok")) throw new IllegalStateException(response.optString("error", "留言发布失败"));
        return messagesFromResponse(response);
    }

    public Map<String, PostMetric> fetchPostMetrics() throws Exception {
        JSONObject response = new JSONObject(request("GET", WORKER_BASE_URL + "/api/post-metrics", null, null));
        if (!response.optBoolean("ok")) throw new IllegalStateException(response.optString("error", "读取文章指标失败"));
        return metricsFromResponse(response);
    }

    public Map<String, PostMetric> recordPostView(String slug) throws Exception {
        JSONObject body = new JSONObject();
        body.put("slug", slug);
        JSONObject response = new JSONObject(request("POST", WORKER_BASE_URL + "/api/post-view", body, null));
        if (!response.optBoolean("ok")) throw new IllegalStateException(response.optString("error", "点击记录失败"));
        return metricsFromResponse(response);
    }

    public Map<String, PostMetric> likePost(String slug, String visitorId) throws Exception {
        JSONObject body = new JSONObject();
        body.put("slug", slug);
        body.put("visitorId", visitorId);
        JSONObject response = new JSONObject(request("POST", WORKER_BASE_URL + "/api/post-like", body, null));
        if (!response.optBoolean("ok")) throw new IllegalStateException(response.optString("error", "点赞失败"));
        return metricsFromResponse(response);
    }

    public void saveBlogCache(JSONObject data) {
        cacheStore.saveBlog(data);
    }

    private VisitStats statsFromResponse(JSONObject response) {
        JSONObject data = response.optJSONObject("data");
        VisitStats stats = new VisitStats();
        if (data != null) {
            stats.visits = data.optInt("visits", 0);
            stats.visitors = data.optInt("visitors", 0);
            stats.lastVisitAt = data.optString("lastVisitAt", "");
        }
        return stats;
    }

    private List<GuestMessage> messagesFromResponse(JSONObject response) {
        List<GuestMessage> messages = new ArrayList<>();
        JSONArray array = response.optJSONArray("data");
        if (array == null) return messages;
        for (int i = 0; i < array.length(); i++) {
            JSONObject item = array.optJSONObject(i);
            if (item == null) continue;
            GuestMessage message = new GuestMessage();
            message.id = item.optString("id", "");
            message.name = item.optString("name", "陌生朋友");
            message.message = item.optString("message", "");
            message.createdAt = item.optString("createdAt", "");
            messages.add(message);
        }
        return messages;
    }

    private Map<String, PostMetric> metricsFromResponse(JSONObject response) {
        Map<String, PostMetric> metrics = new HashMap<>();
        JSONObject data = response.optJSONObject("data");
        if (data == null) return metrics;
        JSONArray names = data.names();
        if (names == null) return metrics;
        for (int i = 0; i < names.length(); i++) {
            String slug = names.optString(i, "");
            JSONObject item = data.optJSONObject(slug);
            if (slug.isEmpty() || item == null) continue;
            PostMetric metric = new PostMetric();
            metric.views = item.optInt("views", 0);
            metric.likes = item.optInt("likes", 0);
            metrics.put(slug, metric);
        }
        return metrics;
    }

    public AdminSession login(String username, String password) throws Exception {
        JSONObject body = new JSONObject();
        body.put("username", username);
        body.put("password", password);
        JSONObject response = new JSONObject(request("POST", WORKER_BASE_URL + "/api/login", body, null));
        if (!response.optBoolean("ok")) throw new IllegalStateException(response.optString("error", "登录失败"));
        AdminSession session = new AdminSession();
        session.user = response.optString("user", username);
        session.token = response.optString("sessionToken", "");
        return session;
    }

    public BlogData fetchAdminData(String token) throws Exception {
        JSONObject response = new JSONObject(request("GET", WORKER_BASE_URL + "/api/posts", null, token));
        if (!response.optBoolean("ok")) throw new IllegalStateException(response.optString("error", "读取后台数据失败"));
        return JsonMapper.parseBlog(response.getJSONObject("data"));
    }

    public DeployResult publish(String token, JSONObject data) throws Exception {
        JSONObject body = new JSONObject();
        body.put("data", data);
        JSONObject response = new JSONObject(request("PUT", WORKER_BASE_URL + "/api/posts", body, token));
        if (!response.optBoolean("ok")) throw new IllegalStateException(response.optString("error", "发布失败"));
        JSONObject deploy = response.optJSONObject("deploy");
        DeployResult result = new DeployResult();
        if (deploy != null) {
            result.commitSha = deploy.optString("commitSha", "");
            result.workflowTriggered = deploy.optBoolean("workflowTriggered", false);
            result.workflowError = deploy.optString("workflowError", "");
            result.message = deploy.optString("message", "");
        }
        return result;
    }

    private String request(String method, String url, JSONObject body, String bearerToken) throws Exception {
        HttpURLConnection connection = null;
        try {
            connection = (HttpURLConnection) new URL(url).openConnection();
            connection.setRequestMethod(method);
            connection.setConnectTimeout(8000);
            connection.setReadTimeout(12000);
            connection.setRequestProperty("Accept", "application/json");
            if (bearerToken != null && !bearerToken.isEmpty()) {
                connection.setRequestProperty("Authorization", "Bearer " + bearerToken);
            }
            if (body != null) {
                connection.setDoOutput(true);
                connection.setRequestProperty("Content-Type", "application/json; charset=utf-8");
                OutputStream output = connection.getOutputStream();
                BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(output, StandardCharsets.UTF_8));
                writer.write(body.toString());
                writer.flush();
                writer.close();
            }

            int code = connection.getResponseCode();
            InputStream stream = code >= 200 && code < 300 ? connection.getInputStream() : connection.getErrorStream();
            String response = readAll(stream);
            if (code < 200 || code >= 300) {
                String message = response;
                try {
                    message = new JSONObject(response).optString("error", response);
                } catch (Exception ignored) {
                }
                throw new IllegalStateException(message);
            }
            return response;
        } catch (SocketTimeoutException | ConnectException | UnknownHostException error) {
            throw new IllegalStateException("无法连接后台 API，请确认当前网络能访问 Cloudflare Worker：" + baseOf(url), error);
        } finally {
            if (connection != null) connection.disconnect();
        }
    }

    private String baseOf(String url) {
        try {
            URL parsed = new URL(url);
            return parsed.getProtocol() + "://" + parsed.getHost();
        } catch (Exception ignored) {
            return url;
        }
    }

    private String readAll(InputStream stream) throws Exception {
        if (stream == null) return "";
        BufferedReader reader = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8));
        StringBuilder builder = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            builder.append(line).append('\n');
        }
        reader.close();
        return builder.toString();
    }
}

File: RaIosNotesApp/RaIosNotesApp/Services/BlogRepository.swift
import Foundation

final class BlogRepository {
    static let publicDataURL = URL(string: "https://galphrui.github.io/ralphrong/data/posts.json")!
    static let workerBaseURL = URL(string: "https://ralphrong-blog-admin.ralphrong.workers.dev")!

    private let cache = CacheStore()
    private let session: URLSession
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()

    init(session: URLSession = .shared) {
        self.session = session
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
    }

    func fetchPublicData() async throws -> (BlogData, DataSource) {
        do {
            var components = URLComponents(url: Self.publicDataURL, resolvingAgainstBaseURL: false)!
            components.queryItems = [URLQueryItem(name: "t", value: String(Int(Date().timeIntervalSince1970 * 1000)))]
            let data = try await requestData(url: components.url!, method: "GET")
            try cache.saveBlogData(data)
            var blog = try decoder.decode(BlogData.self, from: data)
            blog.posts.sort { $0.date > $1.date }
            return (blog, .online("公网数据"))
        } catch {
            if let data = try? cache.loadBlogData() {
                var blog = try decoder.decode(BlogData.self, from: data)
                blog.posts.sort { $0.date > $1.date }
                return (blog, .cache("本地缓存"))
            }
            let data = try Self.loadBundledOfflineData()
            var blog = try decoder.decode(BlogData.self, from: data)
            blog.posts.sort { $0.date > $1.date }
            return (blog, .offline("内置离线数据"))
        }
    }

    func recordVisit(visitorId: String) async throws -> VisitStats {
        do {
            let response: VisitEnvelope = try await requestJSON(path: "/api/visits", method: "POST", body: ["visitorId": visitorId])
            var stats = response.data ?? VisitStats()
            stats.sourceMessage = "公网统计"
            try cache.saveStats(stats)
            return stats
        } catch {
            if let stats = try? cache.loadStats() { return stats }
            throw error
        }
    }

    func fetchVisitStats() async throws -> VisitStats {
        do {
            let response: VisitEnvelope = try await requestJSON(path: "/api/visits", method: "GET")
            var stats = response.data ?? VisitStats()
            stats.sourceMessage = "公网统计"
            try cache.saveStats(stats)
            return stats
        } catch {
            if let stats = try? cache.loadStats() { return stats }
            throw error
        }
    }

    func fetchMessages(postSlug: String = "") async throws -> [GuestMessage] {
        var path = "/api/messages"
        if !postSlug.isEmpty {
            path += "?postSlug=\(postSlug.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? postSlug)"
        }
        let response: ListEnvelope<GuestMessage> = try await requestJSON(path: path, method: "GET")
        return response.data ?? []
    }

    func createMessage(name: String, message: String, postSlug: String = "") async throws -> [GuestMessage] {
        let body = ["name": name, "message": message, "postSlug": postSlug]
        let response: ListEnvelope<GuestMessage> = try await requestJSON(path: "/api/messages", method: "POST", body: body)
        return response.data ?? []
    }

    func fetchPostMetrics() async throws -> [String: PostMetric] {
        let response: DictionaryEnvelope<PostMetric> = try await requestJSON(path: "/api/post-metrics", method: "GET")
        return response.data ?? [:]
    }

    func recordPostView(slug: String) async throws -> [String: PostMetric] {
        let response: DictionaryEnvelope<PostMetric> = try await requestJSON(path: "/api/post-view", method: "POST", body: ["slug": slug])
        return response.data ?? [:]
    }

    func likePost(slug: String, visitorId: String) async throws -> [String: PostMetric] {
        let response: DictionaryEnvelope<PostMetric> = try await requestJSON(path: "/api/post-like", method: "POST", body: ["slug": slug, "visitorId": visitorId])
        return response.data ?? [:]
    }

    func login(username: String, password: String) async throws -> AdminSession {
        let response: LoginEnvelope = try await requestJSON(path: "/api/login", method: "POST", body: ["username": username, "password": password])
        guard response.ok else { throw RepositoryError.server(response.error ?? "登录失败") }
        return AdminSession(user: response.user ?? username, token: response.sessionToken ?? "")
    }

    func fetchAdminData(token: String) async throws -> (BlogData, [String: Any]) {
        let raw = try await requestJSONObject(path: "/api/posts", method: "GET", token: token)
        guard (raw["ok"] as? Bool) == true, let dataObject = raw["data"] as? [String: Any] else {
            throw RepositoryError.server(raw["error"] as? String ?? "读取后台数据失败")
        }
        let data = try JSONSerialization.data(withJSONObject: dataObject)
        var blog = try decoder.decode(BlogData.self, from: data)
        blog.posts.sort { $0.date > $1.date }
        return (blog, dataObject)
    }

    func publish(token: String, json: [String: Any]) async throws -> DeployResult {
        let raw = try await requestJSONObject(path: "/api/posts", method: "PUT", body: ["data": json], token: token)
        guard (raw["ok"] as? Bool) == true else {
            throw RepositoryError.server(raw["error"] as? String ?? "发布失败")
        }
        if let deploy = raw["deploy"] as? [String: Any] {
            let data = try JSONSerialization.data(withJSONObject: deploy)
            return try decoder.decode(DeployResult.self, from: data)
        }
        return DeployResult()
    }

    func saveBlogCache(json: [String: Any]) throws {
        let data = try JSONSerialization.data(withJSONObject: json)
        try cache.saveBlogData(data)
    }

    private func requestJSON<T: Decodable>(path: String, method: String, body: Encodable? = Optional<String>.none, token: String? = nil) async throws -> T {
        let data = try await requestData(url: url(for: path), method: method, body: body, token: token)
        let envelope = try decoder.decode(ServerOKEnvelope.self, from: data)
        if envelope.ok == false { throw RepositoryError.server(envelope.error ?? "请求失败") }
        return try decoder.decode(T.self, from: data)
    }

    private func requestJSONObject(path: String, method: String, body: [String: Any]? = nil, token: String? = nil) async throws -> [String: Any] {
        let data = try await requestData(url: url(for: path), method: method, rawBody: body, token: token)
        return try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
    }

    private func requestData(url: URL, method: String, body: Encodable? = Optional<String>.none, rawBody: [String: Any]? = nil, token: String? = nil) async throws -> Data {
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.timeoutInterval = 12
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let token, !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let body {
            request.httpBody = try encoder.encode(AnyEncodable(body))
            request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        }
        if let rawBody {
            request.httpBody = try JSONSerialization.data(withJSONObject: rawBody)
            request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        }
        let (data, response) = try await session.data(for: request)
        let code = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(code) else {
            let json = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
            throw RepositoryError.server(json?["error"] as? String ?? String(data: data, encoding: .utf8) ?? "HTTP \(code)")
        }
        return data
    }

    private func url(for path: String) -> URL {
        if path.hasPrefix("http") { return URL(string: path)! }
        if path.contains("?") {
            return URL(string: Self.workerBaseURL.absoluteString + path)!
        }
        return Self.workerBaseURL.appendingPathComponent(path.trimmingCharacters(in: CharacterSet(charactersIn: "/")))
    }

    private static func loadBundledOfflineData() throws -> Data {
        guard let url = Bundle.main.url(forResource: "offline-posts", withExtension: "json") else {
            throw RepositoryError.server("找不到内置离线数据")
        }
        return try Data(contentsOf: url)
    }
}

enum DataSource: Equatable {
    case online(String)
    case cache(String)
    case offline(String)

    var message: String {
        switch self {
        case .online(let value), .cache(let value), .offline(let value): return value
        }
    }

    var isReadOnly: Bool {
        if case .online = self { return false }
        return true
    }
}

enum RepositoryError: LocalizedError {
    case server(String)

    var errorDescription: String? {
        switch self {
        case .server(let message): return message
        }
    }
}

private struct ServerOKEnvelope: Decodable {
    var ok: Bool?
    var error: String?
}

private struct VisitEnvelope: Decodable {
    var ok: Bool?
    var error: String?
    var data: VisitStats?
}

private struct ListEnvelope<T: Decodable>: Decodable {
    var ok: Bool?
    var error: String?
    var data: [T]?
}

private struct DictionaryEnvelope<T: Decodable>: Decodable {
    var ok: Bool?
    var error: String?
    var data: [String: T]?
}

private struct LoginEnvelope: Decodable {
    var ok: Bool
    var error: String?
    var user: String?
    var sessionToken: String?
}

private struct AnyEncodable: Encodable {
    private let encodeClosure: (Encoder) throws -> Void

    init(_ wrapped: Encodable) {
        encodeClosure = wrapped.encode
    }

    func encode(to encoder: Encoder) throws {
        try encodeClosure(encoder)
    }
}
